### DISCUSSION TITLE
Transformed geometry - possible performance issue

### DISCUSSION THREAD

**Discussion by AndrejBunjac** [AvaloniaUI/Avalonia#7429] [Q&A]
Ok, so this is a strange one. I am dev working in a company that's using a custom implementation of Avalonia for a game engine. We replaced interfaces such as IDrawingContextImpl and IPlatformRenderInterface with our custom ones so that the rendering goes through our engine's rendering capabilities.

Now the issue surfaced at a time when clipping was broken in our renderer so PushClip method from the IDrawingContextImpl wasn't working as intended. I noticed that when I click on a checkbox - for a few frames I see a huge "tick mark" rendered all over the screen before it disappears and becomes a regular small one contained in the checkbox.

If I understand how this operates, a IStreamGeometryImpl is first going to be created and then the method WithTransform from IGeometry will be called to turn it into a ITransformedGeometryImpl, but what's confusing to me is why is the Stream Geometry ever sent to rendering if it needs to be transformed to its proper size first?

The issue is only visible when no proper clipping is implemented. I am concerned about unnecessary draw calls being sent to renderer for possible performance issue so I wanted to consult with any of the devs about what might be the issue here.

I feel that it's either designed to work like that or I am missing something, although I am having all "not implemented" methods throw. And afaik, I did not implement anything that decides on when are geometries rendered or sent to drawing context for rendering.

So - in short - the question is, does anyone know how the process for creating a transformed geometry works and is the original geometry ever sent to rendering before it actually gets transformed (but outside of clipping bounds).

The relevant interfaces I have custom implementations for are:
IPlatformRenderInterface
IRenderLoop
IRenderTarget
IDrawingContextImpl
IGeometryImpl (and all subgeometries including IStreamGeometryImpl and ITransformedGeometryImpl)

all geometries are implemented as Bezier curve paths and.

Any thoughts were should I look for the cause of this?

**Comment by kekekeks** [comment]
Are you using DeferredRenderer, ImmediateRenderer or a custom implementation?

**Reply by AndrejBunjac** [reply]
In the current state of implementation I'm using ImmediateRenderer - plan to switch to DeferredRenderer later.
